### ISSUE TITLE
quotient_graph seems particularly slow

### ISSUE THREAD

**Issue by dschult** [networkx/networkx#4935]
Given the [speed-up seen in bypassing quotient_graph](https://github.com/networkx/networkx/pull/4929#discussion_r659926874) in #4929 I think we should revisit the code in `quotient_graph`.   And probably in the whole algorithms/minors subpackage.  Maybe it should be turned back into a minors.py module as well...

**Comment by dschult** [comment]
Indeed, the complexity of the quotient graph code is much higher than required for this operation.

Currently the code loops over every pair of sets `b` and `c` of the partition. It checks each pair of nodes between the sets b and c.  If they are connected by an edge, the pairs are no longer checked and an edge is added. But when two partitions are not connected, and in worst case even when they are, it can take a long time to check all node pairs between all partitions.  

A better way to add edges is to go through the edges of G. For each edge, (u, v) connect the partitions that u and v belong to (or add to the weight of the existing edge.  We have to go through all edges of G to find the sum of the edge weights between partitions. It is easy to add the edges during this process and it avoids looking at all the non-edge pairs of nodes.

Rewriting the code with this algorithm should still allow `node-data` `edge-data` to be computed as needed.  The `edge_relation` function will likely have to be made less general.  Currently this function can be any relation and thus doesn’t need to depend on G at all.  That doesn’t really fit with the idea of a quotient graph however.   The new nodes representing partitions should be connected based on connectivity of the original graph.  We could probably get rid of this function argument all together, but maybe there are some use cases we can try to maintan — without having to consider every pair of nodes in every pair of partitions.

**Comment by aaronzo** [comment]
Another consideration for tightening the API for an efficiency gain - I wonder in what percentage of usecases, the desired quotient graph has an equivalence relation where the the relation can only be true when a pair (u, v) share an edge. So far, this has been the case every time I've used `quotient_graph`, and means that my `partition` argument is usually a function that starts:
```python
def equiv(u, v) -> bool:
   if G.has_edge(u, v):
      e = G[u][v].get("attribute-of-interest")
      ...
   elif G,has_edge(v,  u):
      e = G[v][u].get("attribute-of-interest")
      ...
```
Or similar. Both comparisons are needed I think because the `equivalence_classes` function doesn't care about direction. It's even clunkier if you want to write logic on the edge attribute.

In cases like this, I think it could be more efficient to iterate over the set of edges, and create the equivalence classes that way. It would probably be possible to create the edges between the classes in the same loop

**Comment by dschult** [comment]
I agree. But a step further would say that we shouldn't even care about the edge attributes. A quotient graph is defined to have the equivalence classes as nodes and those nodes have an edge iff any node in one equivalence class is connected to any node in the other equivalence class. Should we restrict it this much?  What is the right level of restriction?

Can you say a little bit about how you use the quotient graph and how the `equiv()` functions depend on the edge data from the underlying graph?

**Comment by aaronzo** [comment]
I've used it predominantly in entity merging- for example if I have a graph with business ownerships and transactions and I want to see the connections between the 'brands' or conglomerates rather than each subsidiary - or a similar example where you merge family members into one node to analyze their other links.

**Comment by aaronzo** [comment]
Okay, I've started work on this and #5853 - my proposal is to do the following:
* Write a new `_quotient_graph` function that adds edges to the quotient graph in O(|E|), by looping over original edges. This version of the function will not have an `edge_relation` parameter - it will be a more efficient version of the current default.
* In the user API `quotient_graph` function, either, keep the `edge_relation` parameter - and if it is not `None`, execute the old function.
* add a `self_edges=False` kwarg to `quotient_graph` to give the option to keep edges as per #5853. Keep the default as False to give consistency for existing code.
* I'd also like to change the default behaviour when `node_data=None`. Currently, the default behaviour calculates some [fairly expensive stats](https://github.com/networkx/networkx/blob/c7218d53caa7c8490de89a0950cf539825e6109a/networkx/algorithms/minors/contraction.py#L345-L352) and this is probably unwanted in cases where people just want the graph structure. The new default would be to have no node data in the quotient.
* create a new function, `contract_edges` which allows multiple edges to be contracted in one batch. This will be more efficient than `quotient_graph` in the cases I [described above](https://github.com/networkx/networkx/issues/4935#issuecomment-1484083777).

**Comment by dschult** [comment]
I don't know if this will help or just slow you down, but the function `nx.relabel_nodes` can do the graph construction pretty easily if you make a dict keyed by current node to the equivalence class.  `mapping = {1: "a", 2: "a", 3: "a", 4: "b", 5: "a", 6: "b"}` contracts the nodes {1, 2, 3, 5} into "a" and {4, 6} into "b".  The resulting selfloops would have to be cleaned up. There are of course speed considerations with creating the mapping dict and whatever overhead there is inside the relabel function.

Your description of the new API makes sense to me. :)   Thanks!

**Comment by aaronzo** [comment]
I'm mostly done with the above in #6557 (except for `contract_edges`, which I'll save for a separate PR I think) - but, running benchmarks the new implementation is slower than the old one - and this is mainly down to the `edge_data` parameter.

This parameter takes the form of a function on 2 sets and returns an attribute dictionary: $(A, B) \to dict$, which means that the quotient function has to check every pair in the product of the two sets for existence as an edge $\in G$, before extracting.

What @dschult described [above](https://github.com/networkx/networkx/issues/4935#issuecomment-870246035) would be more suited to a `reduce` style function here, one that accumulates over the edges. The default can still be a sum, but other usecases might be to keep updating a dict, or concatenating a string.

This is a non-backwards compatible change in to the API - but I do think it's worth it. Happy to hear thoughts on the reduce function idea.
